pennylane/grouping/transformations.py

```diff
@@ -135,18 +135,38 @@ def diagonalize_qwc_pauli_words(qwc_grouping):
       PauliZ(wires=[0]) @ PauliZ(wires=[3]),
       PauliZ(wires=[1]) @ PauliZ(wires=[3])])
     """
-    m_paulis = len(qwc_grouping)
-    all_wires = Wires.all_wires([pauli_word.wires for pauli_word in qwc_grouping])
-    wire_map = {label: ind for ind, label in enumerate(all_wires)}
-    for i in range(m_paulis):
-        for j in range(i + 1, m_paulis):
-            if not is_qwc(
-                pauli_to_binary(qwc_grouping[i], wire_map=wire_map),
-                pauli_to_binary(qwc_grouping[j], wire_map=wire_map),
-            ):
-                raise ValueError(
-                    f"{qwc_grouping[i]} and {qwc_grouping[j]} are not qubit-wise commuting."
-                )
+    # If all Paulis are the same kind they are naturally QWC and we
+    # don't need to do the expensive quadratic scalig check
+    first_pauli_name = qwc_grouping[0].name
+    if isinstance(first_pauli_name, list):
+        first_pauli_name = first_pauli_name[0]
+    all_paulis_same_kind = all(
+        all(first_pauli_name == sub_word_name for sub_word_name in pauli_word.name)
+        if isinstance(pauli_word.name, list)
+        else first_pauli_name == pauli_word.name
+        for pauli_word in qwc_grouping
+    )
```

Comment:
I think the logic here might be a bit flawed. One thing to keep in mind is that the each tensor observable does not re-order its name attribute based on the wire order. For this reason the follow two tensor observables will have the same `obs.name`:

```
>>> obs1 = qml.PauliZ(wires=0) @ qml.PauliX(wires=1)
>>> obs2 = qml.PauliZ(wires=1) @ qml.PauiliX(wires=0) 
>>> 
>>> obs1.name == obs2.name 
True
``` 

Since these two are not QWC they should raise an error. But it will get skipped because of this initial check you have added. 

One potential fix for this would be to add a `wire_order` kwarg and reorder manually. I think the performance boost is worth it, but I don't know how often we will come across such a case where all elements in the `qwc_grouping` are the same.

Reply:
Ah! Good catch! 
Can you add a unit test that checks this (such unti test should anyway be in the test set) and then I can fix the implementation.

Reply:
for sure I can 👍🏼